File: src/main/java/com/elchologamer/userlogin/util/command/CommandHandler.java
package com.elchologamer.userlogin.util.command;

import com.elchologamer.userlogin.UserLogin;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;

import java.util.ArrayList;
import java.util.List;

public class CommandHandler extends BaseCommand {

    private final UserLogin plugin = UserLogin.getPlugin();
    private final List<SubCommand> subCommands = new ArrayList<>();

    public CommandHandler(String name) {
        super(name);
    }


    @Override
    public boolean execute(CommandSender sender, Command command, String label, String[] args) {
        if (args.length == 0) return false;

        for (SubCommand subCommand : subCommands) {
            if (!subCommand.getName().equals(args[0])) continue;

            // Check that player has permission
            if (!sender.hasPermission(subCommand.getPermission())) {
                sender.sendMessage(plugin.getMessage("commands.errors.no_permission"));
                return true;
            }

            return subCommand.onCommand(sender, command, label, getSubArgs(args));
        }

        return false;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        List<String> options = new ArrayList<>();

        if (args.length == 1) {
            for (SubCommand subCommand : subCommands) {
                options.add(subCommand.getName());
            }
        } else {
            for (SubCommand subCommand : subCommands) {
                if (!subCommand.getName().equals(args[0])) continue;

                options = subCommand.onTabComplete(sender, command, label, getSubArgs(args));
                break;
            }
        }

        // Filter out list
        if (options != null) options.removeIf(s -> !s.startsWith(args[args.length - 1]));

        return options;
    }

    private String[] getSubArgs(String[] args) {
        String[] subArgs = new String[args.length - 1];
        System.arraycopy(args, 1, subArgs, 0, subArgs.length);
        return subArgs;
    }

    public void add(SubCommand subCommand) {
        subCommands.add(subCommand);
    }
}

File: src/main/kotlin/com/elchologamer/userlogin/command/base/CommandHandler.kt
package com.elchologamer.userlogin.command.base

import com.elchologamer.userlogin.manager.LangManager
import org.bukkit.command.CommandSender

class CommandHandler(name: String) : BaseCommand(name) {
    private val subCommands: MutableList<SubCommand> = ArrayList()

    override fun run(sender: CommandSender, label: String, args: Array<String>): Boolean {
        if (args.isEmpty()) return false

        for (subCommand in subCommands) {
            if (subCommand.name != args[0]) continue

            // Check that player has permission
            if (subCommand.permission != null && !sender.hasPermission(subCommand.permission!!)) {
                sender.sendMessage(LangManager.getMessage("commands.errors.no_permission"))
                return true
            }
            return subCommand.run(sender, label, getSubArgs(args))
        }
        return false
    }

    override fun tabComplete(sender: CommandSender, label: String, args: Array<String>): List<String> {
        val options = if (args.size == 1) {
            subCommands.map { it.name }
        } else {
            subCommands.find { it.name == args[0] }?.tabComplete(sender, label, getSubArgs(args))
        } ?: ArrayList()

        // Filter out list
        return options.filter { it.startsWith(args[args.size - 1]) }
    }

    private fun getSubArgs(args: Array<String>): Array<String> {
        val size = (args.size - 1).coerceAtLeast(0)
        return args.copyOfRange(1.coerceAtMost(size), size)
    }

    fun add(subCommand: SubCommand) {
        subCommands.add(subCommand)
    }
}